zerver/lib/mention.py

```diff
@@ -381,35 +383,37 @@ def init_user_group_data(self, realm_id: int, content: str) -> None:
         if user_group_names_mentions:
             named_user_groups = NamedUserGroup.objects.filter(
                 realm_id=realm_id, name__in=user_group_names_mentions
-            )
+            ).select_related("can_mention_group", "can_mention_group__named_user_group")
 
             # No filter here as we need user_group_name_info for all groups mentions.
             self.user_group_name_info = {group.name.lower(): group for group in named_user_groups}
 
-            # We are interested only in fetching group membersship for
-            # non-deactivated groups and non-silent mentioned groups.
+            # Here we filter the groups so that we ONLY fetch group membership for:
+            # 1- non-deactivated, as deactivated group mentions
+            # are non-operative, so we don't need to waste resources
+            # collecting membership data for them.
+            # 2- non-silent mentioned groups.
+            # 3- Groups the current user has permission to mention,
+            # the fact that any group exists in user_group_names_mentions means the user initially
+            # had the permission to mention that group, but that permission may change
+            # while the message is being sent, in this case we shouldn't fetch that group membership.
             filtered_group_ids = [
                 group.id
                 for group in named_user_groups
                 if not group.deactivated
                 and user_group_names_mentions.get(group.name) == "non-silent"
+                and sender_can_mention_group(self.message_sender, group)
```

Comment:
I don't think this optimization, done this way, is worth it -- you end up doing this check twice for every group we mention, once here and once in `message.py`.